### PR TITLE
#44543 Adjustments to sgtk_from_path for bootstrap

### PR DESCRIPTION
**Pull request by manneohrstrom**
This adjusts the `sgtk_from_path` and `sgtk_from_entity` factory methods to handle bootstrap cases where the pipeline configuration does not have a path registered in Shotgun. Also updates some of the reference documentation.

- Projects no longer need to have a `tank_name` to be discovered by factories.
- Project now needs to be non-archived to be discovered by factories (change in behaviour!)
- Code cleanup and refactor
- Docs
- `TankInitError` exceptions raised instead of `TankErrors`
- Lots and lots of comments.

### CODE REVIEW

File: python/tank/bootstrap/manager.py [hunk 1]
```diff
@@ -627,7 +627,7 @@ def prepare_engine(self, engine_name, entity):
         else:
             log.debug("Configuration has local bundle cache, skipping bundle caching.")
 
-        self._report_progress(self.progress_callback, self._BOOTSTRAP_COMPLETED, "Engine ready.")
+        self._report_progress(self.progress_callback, self._BOOTSTRAP_COMPLETED, "Preparations complete.")
```
Comment (manneohrstrom) [inline python/tank/bootstrap/manager.py]: cut n paste typo that i fixed :)

File: tests/core_tests/test_pipelineconfig_utils.py [hunk 2]
```diff
@@ -29,6 +29,35 @@
...
+            # "   C:/configs\\site//project   "
+            # 1. it has whitespace at the begging and end.
+            # 2. Uses double slashes instead of single backslash
+            # 3. Uses slash instead of backslash
+            # 4. Uses double backslashes as a folder separator.
+            get_install_locations_mock.return_value = sgtk.util.ShotgunPath("   C:/configs\\\\site//project   ", None, None)
+
+            self.assertEqual(
+                # We don't need to pass an actual path since _get_install_location is mocked.
+                tank.pipelineconfig_utils.get_config_install_location(None),
```
[the remark this answers is not in the document]
Comment (manneohrstrom) [reply tests/core_tests/test_pipelineconfig_utils.py]: fixed.

File: python/tank/api.py [hunk 3]
```diff
@@ -798,8 +799,76 @@ def preview_filesystem_structure(self, entity_type, entity_id, engine=None):
 
 def sgtk_from_path(path):
     """
-    Creates a Toolkit Core API instance based on a path inside a project
-    or path pointing directly at a pipeline configuration.
+    Creates a Toolkit Core API instance based on a path to a configuration
```
Comment (manneohrstrom) [inline python/tank/api.py]: note: next up for me is to write docs for bootstrapping - i will do a big pass on the sphinx init docs as part of that, to make sure that we have a decent outline and structure etc.

File: python/tank/pipelineconfig_factory.py [hunk 4]
```diff
@@ -26,155 +25,130 @@
...
-    except TankError:
+    except TankInitError:
         # lookup failed! This may be because there are missing items
         # in the cache. For failures, try again, but this time
         # force re-read the cache (e.g connect to shotgun)
         # if the previous failure was due to a missing item
         # in the cache,
         pc = _from_entity(entity_type, entity_id, force_reread_shotgun_cache=True)
 
+    log.debug("sgtk.from_path resolved %s %s -> %s" % (entity_type, entity_id, pc))
```
Comment (manneohrstrom) [inline python/tank/pipelineconfig_factory.py]: added more debug logging for clarity and to help support.

File: python/tank/pipelineconfig_factory.py [hunk 5]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
+        pipeline_configuration = PipelineConfiguration(config_context_path)
+
+        # get pipeline config shotgun id
+        pc_id = pipeline_configuration.get_shotgun_id()
+
+        # find the pipeline config in our list of configs. If we cannot find it, we
+        # don't have a match beetween the code that is being run and the config
+        # we are trying to start up.
+
+        if pc_id not in [x["id"] for x in all_pc_data]:
```
Comment (manneohrstrom) [inline python/tank/pipelineconfig_factory.py]: this is the actual fix - previously we were comparing by path, but i realised that we could compare by id and all the problems would go away.

File: python/tank/api.py [hunk 6]
```diff
@@ -798,8 +799,76 @@ def preview_filesystem_structure(self, entity_type, entity_id, engine=None):
...
+        import sys
+        sys.path.append("/mnt/toolkit/shared_core")
+
+        # now import the API
+        import sgtk
+
+        # request that the API produced a tk instance suitable for a given file
+        tk = sgtk.sgtk_from_path("/mnt/projects/hidden_forest/shots/aa/aa_001/lighting/foreground.v002.ma")
+
+    .. note::
```
Comment (josh-t) [inline python/tank/api.py]: Wonder if this note should be at the top? That would push people toward the manager before reading through the `sgtk_from_path` specifics.

File: python/tank/errors.py [hunk 7]
```diff
@@ -19,6 +19,12 @@ class TankError(Exception):
     """
 
 
+class TankInitError(TankError):
```
Comment (josh-t) [inline python/tank/errors.py]: Should we avoid creating anything `Tank*`? Maybe call this `ShotgunIntegrationsInitError` or `ToolkitInitError`?

File: python/tank/pipelineconfig_factory.py [hunk 8]
```diff
@@ -26,155 +25,130 @@
...
 
-    :param entity_type: Shotgun Entity type
-    :param entity_id: Shotgun id
-    :param force_reread_shotgun_cache: Should the cache be force re-populated?
-    :returns: Pipeline Configuration object
-    """
+    For info, see :meth:`from_entity`
 
+    :param str entity_type: Shotgun Entity type
+    :param int entity_id: Shotgun id
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: looks like we lost the `force_reread_shotgun_cache` docstring.

File: python/tank/pipelineconfig_factory.py [hunk 9]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
+        - code
+        - windows_path
+        - linux_path
+        - mac_path
+        - project (associated project entity link)
+        - project.Project.tank_name (associated project tank name)
 
-    # extract current os path data from the pipeline configuration shotgun data
-    (all_pc_data, primary_pc_data) = _get_pipeline_configuration_data(associated_sg_pipeline_configs)
+    Thism method will either return a pipeline configuration instance based on
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: This<del>m</del>

File: python/tank/pipelineconfig_factory.py [hunk 10]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
+    one of the ``associated_pipeline_configs`` entries or raise a TankInitError
+    exception.
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: maybe just add the reason why the exception would be raised here? i.e. validation fails.

File: python/tank/pipelineconfig_factory.py [hunk 11]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
 
-    # figure out if we are running a tank command / api from a
-    # local pipeline config or from a studio level install
+    :param list associated_pipeline_configs: Associated Shotgun data.
+    :returns: Pipeline config instance.
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: missing `source` param docstring.

File: python/tank/pipelineconfig_factory.py [hunk 12]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
+                "This error can occur if you have moved a classic toolkit configuration "
+                "manually rather than using the 'tank move_configuration command'. "
+                "It can also occur if you are trying to use a tank command associated "
+                "with one Project to try to operate on a Shot or Asset "
+                "that belongs to another project."
+            )
+
+            raise TankInitError(
+                "You are loading the Toolkit platform from the pipeline configuration "
+                "located in '%s', with Shotgun id %s. You are trying to initializing Toolkit "
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: initialize

File: python/tank/pipelineconfig_factory.py [hunk 13]
```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
...
 
-        if len(primary_pc_data) > 1:
+            raise TankInitError(
+                "%s is associated with more than one Primary pipeline "
+                "configuration. This can happen if there is ambiguity in your project setup, where "
+                "projects store their data in an overlapping fashion, for example if a project is "
+                "named the same as a local storage root. In this case, try creating "
+                "your API instance (or tank command) directly from the pipeline configuration rather "
+                "than via the studio level API. This will explicitly call out which project you are "
+                "intending to use in conjunction with he path. It may also be caused by several projects "
```
Comment (josh-t) [inline python/tank/pipelineconfig_factory.py]: the

File: python/tank/errors.py (hunk 7 above)
Comment (manneohrstrom) [reply python/tank/errors.py]: it's a consistency thing - all things inherited from TankError are TankSomething right now so i just followed that pattern.

**Pull request merged**